[GStreamer][WebRTC] Pipeline revamp in preparation for simulcast support

philn opened the pull request:
#### a5509c9f1bd7777cd6762db9f5ae2538843dd09b
<pre>
[GStreamer][WebRTC] Pipeline revamp in preparation for simulcast support
<a href="https://bugs.webkit.org/show_bug.cgi?id=282926">https://bugs.webkit.org/show_bug.cgi?id=282926</a>

Reviewed by Xabier Rodriguez-Calvar.

Outgoing sources can now multiplex multiple RTP streams, this is going to be required for simulcast
support. Encoding/payloading are now done in a separate sub-bin, managed by new
GStreamer{Audio,Video}RTPPacketizer classes. Encoding RTP parameters validation support was improved
as well.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/fast/mediastream/RTCPeerConnection-inspect-answer-expected.txt:
* LayoutTests/platform/glib/fast/mediastream/RTCPeerConnection-inspect-offer-expected.txt:
* Source/WebCore/Modules/mediastream/MediaStream.cpp:
(WebCore::MediaStream::MediaStream):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
(WebCore::MediaStreamTrack::setMediaStreamId):
(WebCore::MediaStreamTrack::mediaStreamId const):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::initializePipeline):
(WebCore::GStreamerMediaEndpoint::setConfiguration):
(WebCore::getMediaStreamIdsFromSDPMedia):
(WebCore::toGStreamerMediaEndpointTransceiverState):
(WebCore::transceiverStatesFromWebRTCBin):
(WebCore::GStreamerMediaEndpoint::doSetLocalDescription):
(WebCore::GStreamerMediaEndpoint::doSetRemoteDescription):
(WebCore::GStreamerMediaEndpoint::setDescription):
(WebCore::GStreamerMediaEndpoint::processSDPMessage):
(WebCore::GStreamerMediaEndpoint::configureSource):
(WebCore::GStreamerMediaEndpoint::requestPad):
(WebCore::GStreamerMediaEndpoint::addTrack):
(WebCore::GStreamerMediaEndpoint::initiate):
(WebCore::GStreamerMediaEndpoint::trackIdFromSDPMedia):
(WebCore::GStreamerMediaEndpoint::connectIncomingTrack):
(WebCore::GStreamerMediaEndpoint::connectPad):
(WebCore::GStreamerMediaEndpoint::createTransceiverBackends):
(WebCore::GStreamerMediaEndpoint::addTransceiver):
(WebCore::GStreamerMediaEndpoint::createMutedSource):
(WebCore::GStreamerMediaEndpoint::createSourceForTrack):
(WebCore::GStreamerMediaEndpoint::close):
(WebCore::GStreamerMediaEndpoint::suspend):
(WebCore::GStreamerMediaEndpoint::onNegotiationNeeded):
(WebCore::GStreamerMediaEndpoint::collectTransceivers):
(WebCore::GStreamerMediaEndpoint::preprocessStats):
(WebCore::GStreamerMediaEndpoint::configureAndLinkSource): Deleted.
(WebCore::GStreamerMediaEndpoint::createLinkedSourceForTrack): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::addTrack):
(WebCore::GStreamerPeerConnectionBackend::createSourceForTrack):
(WebCore::GStreamerPeerConnectionBackend::existingTransceiverForTrackId):
(WebCore::GStreamerPeerConnectionBackend::createLinkedSourceForTrack): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpReceiverBackend.cpp:
(WebCore::GStreamerRtpReceiverBackend::GStreamerRtpReceiverBackend):
(WebCore::GStreamerRtpReceiverBackend::getParameters):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpReceiverBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpSenderBackend.cpp:
(WebCore::m_initData):
(WebCore::GStreamerRtpSenderBackend::setSource):
(WebCore::GStreamerRtpSenderBackend::replaceTrack):
(WebCore::GStreamerRtpSenderBackend::setParameters):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.cpp:
(WebCore::GStreamerRtpTransceiverBackend::GStreamerRtpTransceiverBackend):
(WebCore::GStreamerRtpTransceiverBackend::createReceiverBackend):
(WebCore::toRtpCodecCapability):
(WebCore::GStreamerRtpTransceiverBackend::setCodecPreferences):
(WebCore::getMsidFromCurrentCodecPreferences): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp:
(WebCore::RTCStatsReport::ReceivedRtpStreamStats::ReceivedRtpStreamStats):
(WebCore::RTCStatsReport::OutboundRtpStreamStats::OutboundRtpStreamStats):
(WebCore::GStreamerStatsCollector::getStats):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::fromRTCEncodingParameters):
(WebCore::toRTCEncodingParameters):
(WebCore::toRTCCodecParameters):
(WebCore::toRTCRtpSendParameters):
(WebCore::fromRTCCodecParameters):
(WebCore::fromRTCSendParameters):
(WebCore::capsFromRtpCapabilities):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:
(WebCore::webrtcKindFromCaps):
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioRTPPacketizer.cpp: Added.
(WebCore::GStreamerAudioRTPPacketizer::create):
(WebCore::GStreamerAudioRTPPacketizer::GStreamerAudioRTPPacketizer):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioRTPPacketizer.h: Copied from Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.h.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerIncomingTrackProcessor.cpp:
(WebCore::GStreamerIncomingTrackProcessor::configure):
(WebCore::GStreamerIncomingTrackProcessor::trackReady):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerRTPPacketizer.cpp: Added.
(WebCore::GStreamerRTPPacketizer::GStreamerRTPPacketizer):
(WebCore::GStreamerRTPPacketizer::configureExtensions):
(WebCore::GStreamerRTPPacketizer::ensureMidExtension):
(WebCore::GStreamerRTPPacketizer::rtpParameters const):
(WebCore::GStreamerRTPPacketizer::rtpStreamId const):
(WebCore::GStreamerRTPPacketizer::payloadType const):
(WebCore::GStreamerRTPPacketizer::currentSequenceNumberOffset const):
(WebCore::GStreamerRTPPacketizer::setSequenceNumberOffset):
(WebCore::GStreamerRTPPacketizer::findLastExtensionId):
(WebCore::GStreamerRTPPacketizer::stats const):
(WebCore::GStreamerRTPPacketizer::startUpdatingStats):
(WebCore::GStreamerRTPPacketizer::updateStatsFromRTPExtensions):
(WebCore::GStreamerRTPPacketizer::stopUpdatingStats):
(WebCore::GStreamerRTPPacketizer::applyEncodingParameters const):
(WebCore::GStreamerRTPPacketizer::reconfigure):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerRTPPacketizer.h: Added.
(WebCore::GStreamerRTPPacketizer::bin const):
(WebCore::GStreamerRTPPacketizer::payloader const):
(WebCore::GStreamerRTPPacketizer::updateStats):
(WebCore::GStreamerRTPPacketizer::configure const):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoRTPPacketizer.cpp: Added.
(WebCore::GStreamerVideoRTPPacketizer::create):
(WebCore::GStreamerVideoRTPPacketizer::GStreamerVideoRTPPacketizer):
(WebCore::GStreamerVideoRTPPacketizer::configure const):
(WebCore::GStreamerVideoRTPPacketizer::updateStats):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoRTPPacketizer.h: Copied from Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCCommon.h.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCCommon.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingAudioSourceGStreamer::RealtimeOutgoingAudioSourceGStreamer):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::initialize):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::linkTee):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::outgoingSourcePad const):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::createPacketizer):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::setPayloadType): Deleted.
(WebCore::RealtimeOutgoingAudioSourceGStreamer::setParameters): Deleted.
(WebCore::RealtimeOutgoingAudioSourceGStreamer::teardown): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::~RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initialize):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::allowedCaps const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::rtpCaps const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::start):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stopOutgoingSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::sourceEnabledChanged):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initializeSourceFromTrackPrivate):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::link):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setSinkPad):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::checkMid):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::parameters):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::codecPreferencesChanged):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::replaceTrack):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setInitialParameters):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::configure):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setParameters):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::getPacketizerForRid):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::linkPacketizer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::configurePacketizers):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::lookupRtpExtensions):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stats):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::startUpdatingStats):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stopUpdatingStats):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::teardown):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::flush): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initializeFromTrack): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::unlinkPayloader): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.h:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setPayloadType): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::fillEncodingParameters): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setParameters): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingVideoSourceGStreamer::RealtimeOutgoingVideoSourceGStreamer):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::initializePreProcessor):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::teardown):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::rtpCapabilities const):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::linkTee):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::outgoingSourcePad const):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::createPacketizer):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::updateStats): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setPayloadType): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::startUpdatingStats): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::stopUpdatingStats): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::sourceEnabledChanged): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::flush): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setParameters): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::fillEncodingParameters): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/286800@main">https://commits.webkit.org/286800@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4677dc5eb51d232236939f044d5d6fed23b169a7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77117 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30032 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81675 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28399 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79234 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65300 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4448 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60427 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18492 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80184 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50389 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/66197 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40734 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47791 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23695 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26722 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68906 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24023 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83102 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4497 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3028 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68709 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4653 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66170 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/67963 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/16951 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10034 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4443 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7259 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4463 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7898 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6222 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -576,7 +579,71 @@ void GStreamerMediaEndpoint::doSetLocalDescription(const RTCSessionDescription*
+
+                auto& sinkPad = source->pad();
+                ASSERT(!gst_pad_is_linked(sinkPad.get()));
+                if (UNLIKELY(gst_pad_is_linked(sinkPad.get()))) {
```
You can put this ASSERT inside the if as ASSERT_NOT_REACHED_WITH_MESSAGE.

philn replied:
Sorry i'm not sure what you're suggesting here :)

calvaris replied:
Remove the ASSERT. Inside the if, do ASSERT_NOT_REACHED[_WITH_MESSAGE].

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1343,28 +1429,100 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
+        g_value_init(&value, GST_TYPE_STRUCTURE);
+        gst_value_set_structure(&value, codecData.get());
+        gst_value_list_append_value(&codecsValue, &value);
+        g_value_unset(&value);
```
If you're unsetting here, wouldn't it be better to append_and_take? Maybe not, I find the GValue API extremely confusing...

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1343,28 +1429,100 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
+            g_value_init(&value, GST_TYPE_STRUCTURE);
+            gst_value_set_structure(&value, encodingData.returnValue().get());
+            gst_value_list_append_value(&encodingsValue, &value);
+            g_value_unset(&value);
```
ditto

cadubentzen reviewed (commented):
Started to look through it but it combines quite a few changes. I think splitting it up would be beneficial.

It seems like a big part of it is extracting code from the outgoing sources into the rtp packetizer classes. A separate refactoring PR with just that without the extra changes would reduce the size of this one quite a bit.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1343,28 +1429,100 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
+        g_value_init(&value, GST_TYPE_STRUCTURE);
+        gst_value_set_structure(&value, encodingData.get());
+        gst_value_list_append_value(&encodingsValue, &value);
+        g_value_unset(&value);
```
ditto

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1613,7 +1769,7 @@ void GStreamerMediaEndpoint::suspend()
         return;
 
     GST_DEBUG_OBJECT(m_pipeline.get(), "Suspending");
-    gst_element_set_state(m_pipeline.get(), GST_STATE_NULL);
+    notImplemented();
```
Why this?

philn replied:
The suspend/resume support needs to be re-assessed after this patch.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -180,32 +210,73 @@ RTCRtpSendParameters toRTCRtpSendParameters(const GstStructure* rtcParameters)
+        GValue value = G_VALUE_INIT;
+        g_value_init(&value, GST_TYPE_STRUCTURE);
+        gst_value_set_structure(&value, codecData.get());
+        gst_value_list_append_value(&codecsValue, &value);
+        g_value_unset(&value);
```
Ditto

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoRTPPacketizer.cpp:
```diff
@@ -0,0 +1,246 @@
+    if (auto maxBitrate = gstStructureGet<unsigned>(encodingParameters, "max-bitrate"_s)) {
+        // maxBitrate is expessed in bits/s but the encoder property is in Kbit/s.
+        if (*maxBitrate >= 1000) {
+            if ((*maxBitrate / 1000) < UINT_MAX)
```
Are this () really necessary?

calvaris reviewed (commented):
I could do just a shallow review because this page is too big.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -576,7 +579,71 @@ void GStreamerMediaEndpoint::doSetLocalDescription(const RTCSessionDescription*
+            if (msid.isEmpty())
+                continue;
+
+            GST_DEBUG_OBJECT(m_pipeline.get(), "Looking-up outgoing source with msid %s", msid.toStringWithoutCopying().ascii().data());
```
The optimal thing, IMHO, it would be to use the `span8` method and get the data from there. Creating a span is cheaper than a String.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1318,10 +1386,28 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
-    auto caps = capsFromRtpCapabilities(m_ssrcGenerator, { .codecs = codecs, .headerExtensions = rtpExtensions }, [&payloadType](GstStructure* structure) {
+    auto msid = emptyString();
+    if (!mediaStreamId.isEmpty() && !trackId.isEmpty())
+        msid = makeString(mediaStreamId, ' ', trackId);
```
I would collapse these three lines in one with ? :

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1343,28 +1429,95 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
+    for (auto& codec : codecs) {
+        auto codecStructure = gst_caps_get_structure(caps.get(), index);
+        auto pt = gstStructureGet<int>(codecStructure, "payload"_s);
+        GUniquePtr<GstStructure> codecData(gst_structure_new("codec-parameters", "pt", G_TYPE_UINT, static_cast<unsigned>(*pt),
```
Don't we need to ensure that *pt is valid?

cadubentzen commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1343,28 +1429,95 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
+        if (allRids.size() == 1 && allRids[0] == emptyString())
+            return Exception { ExceptionCode::TypeError, "Empty rid"_s };
+    } else {
+        GUniquePtr<GstStructure> encodingData(gst_structure_new("encoding-parameters", "encoding-name", G_TYPE_STRING, "H264", "payload", G_TYPE_INT, 97, "active", G_TYPE_BOOLEAN, TRUE, nullptr));
```
Why H264 and payload 97 hardcoded here?

philn replied:
https://datatracker.ietf.org/doc/html/rfc7742#section-6

We could use VP8 too, not sure that's better.

cadubentzen replied:
Hmm iiuc, perhaps VP8 would be better when it gets shipped in distros without proprietary codecs by default.

philn replied:
Well now with openh264 distros are supposed to have some (but bad) h.264 support, or is that a Fedora-only feature?

Anyway, I guess we can check here if h264 is supported and fallback to vp8.

cadubentzen replied:
Btw I don't think this is a blocker for this PR. It can also be done in a follow-up if you prefer.

philn replied:
```diff
diff --git a/Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp b/Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp
index 2ec98f8fd98c..6f9c60f7de5c 100644
--- a/Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp
+++ b/Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp
@@ -1510,11 +1510,23 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
         if (allRids.size() == 1 && allRids[0] == emptyString())
             return Exception { ExceptionCode::TypeError, "Empty rid"_s };
     } else {
-        GUniquePtr<GstStructure> encodingData(gst_structure_new("encoding-parameters", "encoding-name", G_TYPE_STRING, "H264", "payload", G_TYPE_INT, 97, "active", G_TYPE_BOOLEAN, TRUE, nullptr));
-        GValue value = G_VALUE_INIT;
-        g_value_init(&value, GST_TYPE_STRUCTURE);
-        gst_value_set_structure(&value, encodingData.get());
-        gst_value_list_append_and_take_value(&encodingsValue, &value);
+        String fallbackCodec = emptyString();
+        for (auto& codec : codecs) {
+            if (codec.mimeType == "video/H264"_s || codec.mimeType == "video/VP8"_s) {
+                fallbackCodec = codec.mimeType.substring(6);
+                break;
+            }
+        }
+
+        GST_DEBUG_OBJECT(m_pipeline.get(), "Fallback codec: %s", fallbackCodec.ascii().data());
+        if (!fallbackCodec.isEmpty()) {
+            GUniquePtr<GstStructure> encodingData(gst_structure_new("encoding-parameters", "encoding-name", G_TYPE_STRING, fallbackCodec.ascii().data(), "payload", G_TYPE_INT, 97, "active", G_TYPE_BOOLEAN, TRUE, nullptr));
+            GValue value = G_VALUE_INIT;
+            g_value_init(&value, GST_TYPE_STRUCTURE);
+            gst_value_set_structure(&value, encodingData.get());
+            gst_value_list_append_and_take_value(&encodingsValue, &value);
+        } else
+            WTFLogAlways("Missing video encoder / RTP payloader. Please install an H.264 encoder and/or a VP8 encoder");
     }
 
     gst_structure_set_value(initData.get(), "encodings", &encodingsValue);
```

cadubentzen reviewed (commented):
Partially. Still need to look at the new packetizer classes.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioRTPPacketizer.cpp:
```diff
@@ -0,0 +1,165 @@
+
+        if (auto useInbandFec = gstStructureGetString(structure.get(), "useinbandfec"_s)) {
+            if (useInbandFec == "1"_s)
+                g_object_set(encoder.get(), "inband-fec", true, nullptr);
```
TRUE

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
```diff
@@ -143,7 +179,10 @@ void RealtimeOutgoingMediaSourceGStreamer::stopOutgoingSource()
 
-    webkitMediaStreamSrcSignalEndOfStream(WEBKIT_MEDIA_STREAM_SRC(m_outgoingSource.get()));
+    if (WEBKIT_IS_MEDIA_STREAM_SRC(m_outgoingSource.get()))
+        webkitMediaStreamSrcSignalEndOfStream(WEBKIT_MEDIA_STREAM_SRC(m_outgoingSource.get()));
```
Can't you do a _CAST here?

cadubentzen reviewed (approved):
_unofficial r+ (not a reviewer)_

philn committed:
[GStreamer][WebRTC] Pipeline revamp in preparation for simulcast support
https://bugs.webkit.org/show_bug.cgi?id=282926

Reviewed by Xabier Rodriguez-Calvar.

Outgoing sources can now multiplex multiple RTP streams, this is going to be required for simulcast
support. Encoding/payloading are now done in a separate sub-bin, managed by new
GStreamer{Audio,Video}RTPPacketizer classes. Encoding RTP parameters validation support was improved
as well.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/fast/mediastream/RTCPeerConnection-inspect-answer-expected.txt:
* LayoutTests/platform/glib/fast/mediastream/RTCPeerConnection-inspect-offer-expected.txt:
* Source/WebCore/Modules/mediastream/MediaStream.cpp:
(WebCore::MediaStream::MediaStream):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
(WebCore::MediaStreamTrack::setMediaStreamId):
(WebCore::MediaStreamTrack::mediaStreamId const):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::initializePipeline):
(WebCore::GStreamerMediaEndpoint::setConfiguration):
(WebCore::getMediaStreamIdsFromSDPMedia):
(WebCore::toGStreamerMediaEndpointTransceiverState):
(WebCore::transceiverStatesFromWebRTCBin):
(WebCore::GStreamerMediaEndpoint::doSetLocalDescription):
(WebCore::GStreamerMediaEndpoint::doSetRemoteDescription):
(WebCore::GStreamerMediaEndpoint::setDescription):
(WebCore::GStreamerMediaEndpoint::processSDPMessage):
(WebCore::GStreamerMediaEndpoint::configureSource):
(WebCore::GStreamerMediaEndpoint::requestPad):
(WebCore::GStreamerMediaEndpoint::addTrack):
(WebCore::GStreamerMediaEndpoint::initiate):
(WebCore::GStreamerMediaEndpoint::trackIdFromSDPMedia):
(WebCore::GStreamerMediaEndpoint::connectIncomingTrack):
(WebCore::GStreamerMediaEndpoint::connectPad):
(WebCore::GStreamerMediaEndpoint::createTransceiverBackends):
(WebCore::GStreamerMediaEndpoint::addTransceiver):
(WebCore::GStreamerMediaEndpoint::createMutedSource):
(WebCore::GStreamerMediaEndpoint::createSourceForTrack):
(WebCore::GStreamerMediaEndpoint::close):
(WebCore::GStreamerMediaEndpoint::suspend):
(WebCore::GStreamerMediaEndpoint::onNegotiationNeeded):
(WebCore::GStreamerMediaEndpoint::collectTransceivers):
(WebCore::GStreamerMediaEndpoint::preprocessStats):
(WebCore::GStreamerMediaEndpoint::configureAndLinkSource): Deleted.
(WebCore::GStreamerMediaEndpoint::createLinkedSourceForTrack): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::addTrack):
(WebCore::GStreamerPeerConnectionBackend::createSourceForTrack):
(WebCore::GStreamerPeerConnectionBackend::existingTransceiverForTrackId):
(WebCore::GStreamerPeerConnectionBackend::createLinkedSourceForTrack): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpReceiverBackend.cpp:
(WebCore::GStreamerRtpReceiverBackend::GStreamerRtpReceiverBackend):
(WebCore::GStreamerRtpReceiverBackend::getParameters):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpReceiverBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpSenderBackend.cpp:
(WebCore::m_initData):
(WebCore::GStreamerRtpSenderBackend::setSource):
(WebCore::GStreamerRtpSenderBackend::replaceTrack):
(WebCore::GStreamerRtpSenderBackend::setParameters):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.cpp:
(WebCore::GStreamerRtpTransceiverBackend::GStreamerRtpTransceiverBackend):
(WebCore::GStreamerRtpTransceiverBackend::createReceiverBackend):
(WebCore::toRtpCodecCapability):
(WebCore::GStreamerRtpTransceiverBackend::setCodecPreferences):
(WebCore::getMsidFromCurrentCodecPreferences): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp:
(WebCore::RTCStatsReport::ReceivedRtpStreamStats::ReceivedRtpStreamStats):
(WebCore::RTCStatsReport::OutboundRtpStreamStats::OutboundRtpStreamStats):
(WebCore::GStreamerStatsCollector::getStats):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::fromRTCEncodingParameters):
(WebCore::toRTCEncodingParameters):
(WebCore::toRTCCodecParameters):
(WebCore::toRTCRtpSendParameters):
(WebCore::fromRTCCodecParameters):
(WebCore::fromRTCSendParameters):
(WebCore::capsFromRtpCapabilities):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:
(WebCore::webrtcKindFromCaps):
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioRTPPacketizer.cpp: Added.
(WebCore::GStreamerAudioRTPPacketizer::create):
(WebCore::GStreamerAudioRTPPacketizer::GStreamerAudioRTPPacketizer):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioRTPPacketizer.h: Copied from Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.h.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerIncomingTrackProcessor.cpp:
(WebCore::GStreamerIncomingTrackProcessor::configure):
(WebCore::GStreamerIncomingTrackProcessor::trackReady):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerRTPPacketizer.cpp: Added.
(WebCore::GStreamerRTPPacketizer::GStreamerRTPPacketizer):
(WebCore::GStreamerRTPPacketizer::configureExtensions):
(WebCore::GStreamerRTPPacketizer::ensureMidExtension):
(WebCore::GStreamerRTPPacketizer::rtpParameters const):
(WebCore::GStreamerRTPPacketizer::rtpStreamId const):
(WebCore::GStreamerRTPPacketizer::payloadType const):
(WebCore::GStreamerRTPPacketizer::currentSequenceNumberOffset const):
(WebCore::GStreamerRTPPacketizer::setSequenceNumberOffset):
(WebCore::GStreamerRTPPacketizer::findLastExtensionId):
(WebCore::GStreamerRTPPacketizer::stats const):
(WebCore::GStreamerRTPPacketizer::startUpdatingStats):
(WebCore::GStreamerRTPPacketizer::updateStatsFromRTPExtensions):
(WebCore::GStreamerRTPPacketizer::stopUpdatingStats):
(WebCore::GStreamerRTPPacketizer::applyEncodingParameters const):
(WebCore::GStreamerRTPPacketizer::reconfigure):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerRTPPacketizer.h: Added.
(WebCore::GStreamerRTPPacketizer::bin const):
(WebCore::GStreamerRTPPacketizer::payloader const):
(WebCore::GStreamerRTPPacketizer::updateStats):
(WebCore::GStreamerRTPPacketizer::configure const):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoRTPPacketizer.cpp: Added.
(WebCore::GStreamerVideoRTPPacketizer::create):
(WebCore::GStreamerVideoRTPPacketizer::GStreamerVideoRTPPacketizer):
(WebCore::GStreamerVideoRTPPacketizer::configure const):
(WebCore::GStreamerVideoRTPPacketizer::updateStats):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoRTPPacketizer.h: Copied from Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCCommon.h.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCCommon.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingAudioSourceGStreamer::RealtimeOutgoingAudioSourceGStreamer):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::initialize):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::linkTee):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::outgoingSourcePad const):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::createPacketizer):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::setPayloadType): Deleted.
(WebCore::RealtimeOutgoingAudioSourceGStreamer::setParameters): Deleted.
(WebCore::RealtimeOutgoingAudioSourceGStreamer::teardown): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::~RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initialize):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::allowedCaps const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::rtpCaps const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::start):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stopOutgoingSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::sourceEnabledChanged):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initializeSourceFromTrackPrivate):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::link):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setSinkPad):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::checkMid):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::parameters):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::codecPreferencesChanged):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::replaceTrack):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setInitialParameters):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::configure):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setParameters):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::getPacketizerForRid):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::linkPacketizer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::configurePacketizers):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::lookupRtpExtensions):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stats):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::startUpdatingStats):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stopUpdatingStats):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::teardown):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::flush): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initializeFromTrack): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::unlinkPayloader): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.h:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setPayloadType): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::fillEncodingParameters): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setParameters): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingVideoSourceGStreamer::RealtimeOutgoingVideoSourceGStreamer):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::initializePreProcessor):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::teardown):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::rtpCapabilities const):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::linkTee):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::outgoingSourcePad const):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::createPacketizer):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::updateStats): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setPayloadType): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::startUpdatingStats): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::stopUpdatingStats): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::sourceEnabledChanged): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::flush): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setParameters): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::fillEncodingParameters): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.h:

Canonical link: https://commits.webkit.org/286800@main
